cmd/pipecd/ops.go

```diff
@@ -79,6 +80,35 @@ func (s *ops) run(ctx context.Context, t cli.Telemetry) error {
 		return err
 	}
 
+	// Prepare sql database.
+	if cfg.Datastore.Type == model.DataStoreMySQL {
+		mysqlEnsurer := mysqlensurer.NewMySQLEnsurer(
+			cfg.Datastore.MySQLConfig.URL,
+			cfg.Datastore.MySQLConfig.Database,
+			cfg.Datastore.MySQLConfig.UsernameFile,
+			cfg.Datastore.MySQLConfig.PasswordFile,
+			t.Logger,
+		)
+		err := mysqlEnsurer.EnsureSchema(ctx)
+		if err != nil {
+			t.Logger.Error("failed to prepare sql database", zap.Error(err))
+			return err
+		}
+		// No need to run this create indexes operation in routine because it runs asynchronously.
+		// ref: https://dev.mysql.com/doc/refman/8.0/en/innodb-online-ddl-operations.html#online-ddl-index-operations
+		err = mysqlEnsurer.EnsureIndexes(ctx)
+		if err != nil {
+			t.Logger.Error("failed to create required indexes on sql database", zap.Error(err))
+			return err
+		}
+		defer func() {
+			// Close connection held by the client.
+			if err := mysqlEnsurer.Close(); err != nil {
+				t.Logger.Error("failed to close database ensurer connection", zap.Error(err))
+			}
+		}()
```

Comment:
There you go. Paying attention to close connection is great. But I would say this should be defined right after calling `mysqlensurer.NewMySQLEnsurer()`. If it goes on like this, it couldn't be closed in some cases.

Reply:
And I feel connect() should be called in the `NewMySQLEnsurer()` method, but are there any reasons you don't do so?

Reply:
I get your point 😂 Thanks for pointing out 🙏 Addressed on https://github.com/pipe-cd/pipe/pull/1699/commits/b320ac325c82ecf04e827feefe54bf1d73a4f69a

Reply:
> And I feel connect() should be called in the `NewMySQLEnsurer()` method, but are there any reasons you don't do so?

Just don't want to make a connection to the remote database until it has to be 😁 In the viewpoint of `ops`, one we make MySQLEnsurer instance, we know we will make call to `EnsureSchema` and `EnsureIndexes` but from the viewpoint of `MySQLEnsurer` itself, we could reduce cost by making the connection on required only.

Reply:
I hear you. But I feel like panic can occur when calling Close() if nil was populated to the client, but am I wrong? 😃

Reply:
You guess right 👍 Let's add check `nil` for the client before close 😂

Reply:
I think it would be better to get the connection on the caller when it is created (NewMySQLEnsurer), and close it on the caller when it is finished to use.

>but from the viewpoint of MySQLEnsurer itself, we could reduce cost by making the connection on required only.

I think it would be less costly to generate the connections first and then use them around, but what do you think?

Reply:
>You guess right 👍 Let's add check nil for the client before close 😂

No, no. I didn't mean that. I was just trying to say we should get the connection in the `NewMySQLEnsurer()` anyway.

Reply:
Ah, I was overlooked it checks if the client is already set and if so it re-use it.

```
	if m.client != nil {
		return nil
	}
```

Reply:
>Just don't want to make a connection to the remote database until it has to be 

Okay, I got your point. Neve mind 😄

Reply:
>You guess right 👍 Let's add check nil for the client before close 😂

Sorry, if it goes on like that, please add this check just in case 🙏

Reply:
To be honest, I know we could make connection to the remote database in `NewMySQLEnsurer` as we do in `XXXDatastore` from other parts of PipeCD source code (just as your suggestion). Just my thought, but I think for this task (ensure SQL schema/indexes), I want to reduce the cost of making connections to the remote database by postponing it until the time we have to make it. And the downside was we have to check nullish of the `m.client` each time we make use of it 😂 
Let's me add check null before close as mentioned on https://github.com/pipe-cd/pipe/pull/1699#discussion_r594060843 🙏

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1699/commits/e57d0a1dce584bf3a657cdf0ba188ed7005655c3 🙏

Reply:
👍  Ok. I am ok with adding the check inside the Close function.
Yes, we should not add any network dealings inside the `New...` function. Let `New...` returns immediately.

Reply:
Okay, that makes total sense. I appreciate you taking the time to explain this to me!

Reply:
Thank you both 😁

Reply:
But I have just checked the doc again. It says

> Open may just validate its arguments without creating a connection to the database. To verify that the data source name is valid, call Ping.

https://github.com/go-sql-driver/mysql/wiki/Examples

Actually, `db.Open` will not make any network dealings.
So we can rename the `connect` function to `createClient` and call it inside `New`.

Reply:
🤣 so I just overthinking and redo lib works 😂 Let's me commit that 👌

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1699/commits/027692c13f2b610e020b5bfd3ce79cf8ad2ba931 PTAL 🙏